File: benchmarks/fixtures/webapp_rb/middleware/cors.rb
# CORS middleware.

require_relative '../utils/helpers'

DEFAULT_ORIGINS = ['http://localhost:3000', 'https://app.example.com'].freeze

# CORS policy configuration.
class CorsPolicy
  attr_accessor :allowed_origins, :allowed_methods, :allow_credentials, :max_age

  def initialize(
    allowed_origins: DEFAULT_ORIGINS,
    allowed_methods: %w[GET POST PUT DELETE],
    allow_credentials: true,
    max_age: 86_400
  )
    @allowed_origins = allowed_origins
    @allowed_methods = allowed_methods
    @allow_credentials = allow_credentials
    @max_age = max_age
  end

  def origin_allowed?(origin)
    @allowed_origins.include?('*') || @allowed_origins.include?(origin)
  end

  def headers(origin)
    return {} unless origin_allowed?(origin)

    {
      'Access-Control-Allow-Origin' => origin,
      'Access-Control-Allow-Methods' => @allowed_methods.join(', '),
      'Access-Control-Max-Age' => @max_age.to_s
    }
  end
end

# Apply CORS headers to a request.
def cors_middleware(request, policy = nil)
  validate_request(request)
  cors = policy || CorsPolicy.new
  origin = request[:origin] || ''
  if origin && !origin.empty?
    hdrs = cors.headers(origin)
    get_logger('middleware.cors').warn("CORS rejected: #{origin}") if hdrs.empty?
    return request.merge(cors_headers: hdrs)
  end
  request.merge(cors_headers: {})
end

File: benchmarks/fixtures/webapp_rs/middleware/cors.rs
use crate::utils::helpers::get_logger;
use crate::Response;

/// Allowed origins for CORS requests.
const ALLOWED_ORIGINS: &[&str] = &["http://localhost:3000", "https://app.example.com"];

/// Allowed HTTP methods for CORS.
const ALLOWED_METHODS: &[&str] = &["GET", "POST", "PUT", "DELETE", "OPTIONS"];

/// Check if an origin is allowed by the CORS policy.
pub fn is_origin_allowed(origin: &str) -> bool {
    let logger = get_logger("middleware.cors");
    let allowed = ALLOWED_ORIGINS.contains(&origin);
    logger.info(&format!("CORS origin check: {} = {}", origin, allowed));
    allowed
}

/// Add CORS headers to a response for the given origin.
pub fn add_cors_headers(response: &mut Response, origin: &str) {
    let logger = get_logger("middleware.cors");
    if is_origin_allowed(origin) {
        logger.info(&format!("Adding CORS headers for origin: {}", origin));
    } else {
        logger.warn(&format!("Rejected CORS origin: {}", origin));
    }
}

/// Handle a CORS preflight OPTIONS request.
pub fn handle_preflight(origin: &str) -> Response {
    let logger = get_logger("middleware.cors");
    logger.info(&format!("Handling CORS preflight for: {}", origin));
    if is_origin_allowed(origin) {
        Response::ok("OK".to_string())
    } else {
        Response::error(403, "Origin not allowed")
    }
}
